ensure duplicates are removed before enforcing collations

roperzh opened the pull request:
Related to #10787, this tries to find in the tables with High likelihood described in the issue.

This successfully accounts for unique keys that contain leading/trailing whitespace and are using a collation with a pad attribute set to `NO PAD` (considers whitespace as any other character instead of ignoring it)

I haven't found a way to successfully detect the same scenario for special unicode characters, for example:

```
mysql> SELECT TABLE_NAME, TABLE_COLLATION FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = 'software';
+------------+--------------------+
| TABLE_NAME | TABLE_COLLATION    |
+------------+--------------------+
| software   | utf8mb4_general_ci |
+------------+--------------------+
1 row in set (0.01 sec)

mysql> select vendor COLLATE utf8mb4_unicode_ci from software where name = 'zchunk-libs' GROUP BY vendor COLLATE utf8mb4_unicode_ci;
+-----------------------------------+
| vendor COLLATE utf8mb4_unicode_ci |
+-----------------------------------+
| vendor                            |
| vendor?                           |
+-----------------------------------+
2 rows in set (0.01 sec)

mysql> ALTER TABLE `software` CONVERT TO CHARACTER SET `utf8mb4` COLLATE `utf8mb4_unicode_ci`;
ERROR 1062 (23000): Duplicate entry 'zchunk-libs-1.2.1-rpm_packages--vendor\2007-x86_64' for key 'unq_name'
```
> **Note** that `?`  in "vendor?" is an unicode character


# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [ ] Changes file added for user-visible changes in `changes/` or `orbit/changes/`.
  See [Changes files](https://fleetdm.com/docs/contributing/committing-changes#changes-files) for more information.
- [ ] Documented any API changes (docs/Using-Fleet/REST-API.md or docs/Contributing/API-for-contributors.md)
- [ ] Documented any permissions changes
- [ ] Input data is properly validated, `SELECT *` is avoided, SQL injection is prevented (using placeholders for values in statements)
- [ ] Added support on fleet's osquery simulator `cmd/osquery-perf` for new osquery data ingestion features.
- [ ] Added/updated tests
- [ ] Manual QA for all new/changed functionality
  - For Orbit and Fleet Desktop changes:
    - [ ] Manual QA must be performed in the three main OSs, macOS, Windows and Linux.
    - [ ] Auto-update manual QA, from released version of component to new version (see [tools/tuf/test](../tools/tuf/test/README.md)).

roperzh commented on server/datastore/mysql/migrations/tables/20230328091816_EnsureUniformCollation.go:
```diff
@@ -3,14 +3,170 @@ package tables
+
+	defer rows.Close()
+	var idGroups [][]uint
+	for rows.Next() {
```
here and below, I tried to use `sqlx.Tx`, but I was getting a panic while it tried to decode the return value. In the end I gave up and resorted to this.

mna reviewed (approved):
The logic makes sense to me, and there are good tests, LGTM. You may want to add more verbose  logging in the `fixup` functions to get more visibility into what it encountered/how it behaves in case it doesn't fix the issue, if there's something else at play?

Also, I know we always recommend as best practice to make a backup of the DB before upgrading, but it may be worth insisting on this as a refresher for that migration, given that it may delete data (duplicates, but still).

zwass commented on server/datastore/mysql/migrations/tables/20230328091816_EnsureUniformCollation.go:
```diff
@@ -103,14 +270,14 @@ func changeCollation(tx *sql.Tx, charset string, collation string) (err error) {
 }
 
-func Up_20230315104937(tx *sql.Tx) error {
+func Up_20230328091816(tx *sql.Tx) error {
```
What do you think about leaving the migration name the same? I'm imagining that allows installations that have not yet upgraded to use the new improved migration, while installations that already successfully ran the migration will do nothing.

roperzh replied:
I think my first instinct was to treat migrations as immutable, per the contributor guidelines.

Thinking about this, I think you're right, in this scenario that's the ideal for us. I have renamed the migration to its old name, thanks!

zwass commented on server/datastore/mysql/migrations/tables/20230328091816_EnsureUniformCollation.go:
```diff
@@ -3,14 +3,170 @@ package tables
+}
+
+func fixupSoftware(tx *sql.Tx, collation string) error {
+	rows, err := tx.Query(`
```
Can we explicitly disable the `gosec` check for this? Ideally adding a comment like "string formatting must be used here, but input is not user-controllable".

zwass commented on server/datastore/mysql/migrations/tables/20230328091816_EnsureUniformCollation.go:
```diff
@@ -3,14 +3,170 @@ package tables
+			if _, err := tx.Exec("DELETE FROM software_host_counts WHERE software_id = ?", ids[i]); err != nil {
+				return err
+			}
+			if _, err := tx.Exec("UPDATE host_software SET software_id = ? WHERE software_id = ?", ids[0], ids[i]); err != nil {
```
Is it possible for this to generate duplicate `host_software` entries for a given host? If so, any idea whether that could cause problems?

roperzh replied:
this is a good point, I don't foresee any problems, but since these are dupes, it's probably better to `DELETE` them, what do you think?

zwass replied:
I think there are tradeoffs to either approach but I feel slightly better about deleting.

roperzh commented:
> You may want to add more verbose logging in the fixup functions to get more visibility into what it encountered/how it behaves in case it doesn't fix the issue, if there's something else at play?

great point, thanks! I wrapped all errors with useful info, and added an `INFO` log, is there anything else you see that can be improved?

> Also, I know we always recommend as best practice to make a backup of the DB before upgrading, but it may be worth insisting on this as a refresher for that migration, given that it may delete data (duplicates, but still).

great point, I added a note to the changelog entry, I'm thinking we make sure that lands into the release notes, is there anything else we could do?

roperzh commented:
@mna @zwass thanks so much for the reviews. I have addressed the changes and all tests are passing. Would you mind taking another look?

mna commented:
@roperzh 

>  I wrapped all errors with useful info, and added an INFO log, is there anything else you see that can be improved?

That looks good to me, it should give good insights about what it found (or not) if it still fails.

>  I added a note to the changelog entry, I'm thinking we make sure that lands into the release notes, is there anything else we could do?

:+1: sounds good to me.

zwass reviewed (approved):
I think we are good to go now. Even if it doesn't catch every case, it seems like an improvement. Thank you!
